### PR TITLE
Add Reserve Cash

### PR DESCRIPTION
**Pull request by nimishagarwal76**

### CODE REVIEW

File: models/User.go [hunk 1]
```diff
@@ -729,6 +732,26 @@ func SubtractUserCash(user *User, fee uint64, tx *gorm.DB) error {
 	return nil
 }
 
+// AddUserReservedCash subtracts reservedCash for users account
```
Comment (gauthamk97) [inline models/User.go]: Comment is incorrect about what this function does. plz fix.

File: models/User.go [hunk 2]
```diff
@@ -729,6 +732,26 @@ func SubtractUserCash(user *User, fee uint64, tx *gorm.DB) error {
...
+// This function should be called ONLY AFTER lock is obtained on user
+func AddUserReservedCash(user *User, fee uint64, tx *gorm.DB) error {
```
Comment (gauthamk97) [inline models/User.go]: Why are you calling this parameter `fee`? It's not a fee at all. It's cash that's being reserved. Give it a more accurate name.

File: models/User.go [hunk 3]
```diff
@@ -828,13 +851,15 @@ func PlaceAskOrder(userId uint32, ask *Ask) (uint32, error) {
 	l.Debugf("Check3: Passed. Creating Ask.")
 
 	oldCash := user.Cash
+	oldReservedCash := user.ReservedCash
```
Comment (gauthamk97) [inline models/User.go]: Why is this needed? You're not modifying the reservedCash anywhere in this function

File: models/User.go [hunk 4]
```diff
@@ -828,13 +851,15 @@ func PlaceAskOrder(userId uint32, ask *Ask) (uint32, error) {
...
 
 	db := getDB()
 	tx := db.Begin()
 
 	var errorHelper = func(format string, args ...interface{}) (uint32, error) {
 		l.Errorf(format, args...)
 		user.Cash = oldCash
+		user.ReservedCash = oldReservedCash
```
Comment (gauthamk97) [inline models/User.go]: This is also not needed for the same reason mentioned above

File: models/User.go [hunk 5]
```diff
@@ -977,7 +1002,8 @@ func PlaceBidOrder(userId uint32, bid *Bid) (uint32, error) {
 	orderPrice := getOrderFeePrice(bid.Price, bid.StockId, bid.OrderType)
 	orderFee := getOrderFee(bid.StockQuantity, orderPrice)
 	cashLeft := int64(user.Cash) - int64(bid.StockQuantity*orderPrice+orderFee)
-
+	reservedCash := uint64(bid.StockQuantity * orderPrice)
+	l.Debugf("Reserved cash is %d", reservedCash)
```
Comment (gauthamk97) [inline models/User.go]: Rephrase this to "Cash to be reserved is %d"

Comment (gauthamk97) [reply models/User.go]: And add a blank line above this.

File: models/User.go [hunk 6]
```diff
@@ -1009,6 +1037,10 @@ func PlaceBidOrder(userId uint32, bid *Bid) (uint32, error) {
 		return errorHelper("Error while subtracting order fee from user. Rolling back. Error: %+v", err)
 	}
 
+	if err := AddUserReservedCash(user, reservedCash, tx); err != nil {
+		return errorHelper("Error while subtracting reserved cash from user. Rolling back. Error: %+v", err)
```
Comment (gauthamk97) [inline models/User.go]: How is this error message accurate? Please check what you're copy pasting. Update the error message.

File: models/User.go [hunk 7]
```diff
@@ -1132,6 +1164,7 @@ func saveBidCancelOrderTransaction(bidOrder *Bid, user *User, tx *gorm.DB) error
 	)
 
 	user.Cash += uint64(reservedCash)
+	user.ReservedCash -= uint64(reservedCash)
```
Comment (gauthamk97) [inline models/User.go]: `saveBidCancelOrderTransaction` is called by `CancelOrder` in User.go. If an error is thrown by `saveBidCancelOrderTransaction`, a tx.Rollback() is called. However, we are not reverting to the old state.

This was not needed before, because we were not changing the user's cash or reserved cash. But now since we introduced reserving stocks and cash, we need to also revert to old state on top of rolling the database back. 

So CancelOrder() needs to be updated with this.

Comment (gauthamk97) [reply models/User.go]: And also add a blank line after this

File: models/User.go [hunk 8]
```diff
@@ -1764,10 +1797,20 @@ func PerformOrderFillTransaction(ask *Ask, bid *Bid, stockTradePrice uint64, sto
...
 	askingUser.Cash += uint64(stockTradeQty) * stockTradePrice
-	biddingUser.Cash = uint64(cashLeft)
+
+	// bidding user's cash will first be subtracted from reserved cash and then if required from bidding user's cash
+	cashDifference := int64(biddingUser.ReservedCash) - total
+	if cashDifference >= 0 {
+		biddingUser.ReservedCash -= uint64(total)
+	} else {
+		biddingUser.ReservedCash = 0                                        // reserved cash gets all used
+		biddingUser.Cash -= uint64(total - int64(biddingUser.ReservedCash)) // remaining cash is used completely, money taken from cash
```
Comment (gauthamk97) [inline models/User.go]: This line should not be there. It was correct the way it was before (i.e) `biddingUser.Cash = uint64(cashLeft)`.

All you need to do is update `biddingUser.ReservedCash`. And for that it would be better if you used 

```go
biddingUser.ReservedCash = math.Max(0, biddingUser.ReservedCash - total)
```

This is simpler than using if else statements.

File: models/User.go [hunk 9]
```diff
@@ -2095,6 +2140,50 @@ func GetStocksOwned(userId uint32) (map[uint32]int64, error) {
...
+	}
+	l.Debugf("Acquired")
+	defer func() {
+		close(ch)
+		l.Debugf("Released lock on user")
+	}()
+
+	db := getDB()
+
+	sql := "SELECT SUM(stockQuantity), SUM(stockQuantityFulFilled), stockId FROM `Bids` WHERE `userId` = ? AND isClosed = 0 GROUP BY `stockId`"
```
Comment (gauthamk97) [inline models/User.go]: Aren't you supposed to make this query from the Asks table? It's Ask orders that cause stocks to be reserved. 

@PaiAkshay998 Check if this query will work.

Comment (PaiAkshay998) [reply models/User.go]: I guess this should be right. Whatever is not fulfilled is reserved at this point. So this should be fine.

**Comment by Harsh2098** [comment]
Lots of changes to be made. On top of all the comments, I see you're storing the reservedCash as part of the Users model, but is there an API to actually access it? I couldn't find that.


In LoginResponse you'll get User, so there you'll get initial reservedCash.

File: models/User.go (hunk 9 above)
Comment (gauthamk97) [reply models/User.go]: No, but the query is made on the wrong table no. Shouldn't it be from the `Asks` table? Not the Bids table

Comment (nimishagarwal76) [reply models/User.go]: When you buy an order, it gets stored in Bids table, it has ``stockQuantity`` and ``stockQuantityFulFilled``, difference of the fields basically gives the stocks that user has reserved(he paid cash but he did not get).

File: models/User.go (hunk 8 above)
Comment (nimishagarwal76) [reply models/User.go]:
```
It was correct the way it was before (i.e) `biddingUser.Cash = uint64(cashLeft)`.
@gauthamk97 
```
Say if user purchased stocks, but gets it for less, then only reserved cash would be deducted, user cash won't change,
Doing your way would cause user cash to increase also which was required earlier but now not required as there is reserved cash for that.

<hr>

Better way would be

``` biddingUser.ReservedCash = math.Max(0, biddingUser.ReservedCash - total) ```
``` biddingUser.Cash += math.Min(0, biddingUser.ReservedCash - total) ```

File: models/User.go (hunk 9 above)
Comment (gauthamk97) [reply models/User.go]: The user has only reserved stocks if he places a Sell order, hence you should be checking the Asks table.

File: models/User.go [hunk 10]
```diff
@@ -978,6 +1001,8 @@ func PlaceBidOrder(userId uint32, bid *Bid) (uint32, error) {
 	orderFee := getOrderFee(bid.StockQuantity, orderPrice)
 	cashLeft := int64(user.Cash) - int64(bid.StockQuantity*orderPrice+orderFee)
```
Comment (PaiAkshay998) [inline models/User.go]: We're declaring reservedCash variable below. Can we simply use that here instead of doing bid.StockQuantity*orderPrice. Makes it slightly more readable
